src/low-level/imap/mailimap_parser.c

```diff
@@ -6017,43 +6020,51 @@ static int mailimap_literal_parse_progress(mailstream * fd, MMAPString * buffer,
     res = r;
     goto err;
   }
-  
-  literal = mmap_string_sized_new(number);
+
+  if (use_msg_body_handler) {
+    literal = mmap_string_new("");
+  }
+  else {
+    literal = mmap_string_sized_new(number);
+  }
   if (literal == NULL) {
     res = MAILIMAP_ERROR_MEMORY;
     goto err;
   }
   
   left = (uint32_t) (buffer->len - cur_token);
-  
-  if (left >= number) {
-    if (number > 0)
-      if (mmap_string_append_len(literal, buffer->str + cur_token,
-                                 number) == NULL) {
-        res = MAILIMAP_ERROR_MEMORY;
-        goto free_literal;
-      }
-    if (progr_rate != 0) {
-      if (progr_fun != NULL) {
-        progr_fun(number, number);
+  needed = number;
+
+  uint32_t current_prog = 0;
+
+  if (left > 0) {
```

Comment:
Is there a way to rewrite it that would be similar when comparing with the previous code?
I makes it hard for me to figure out what is going to break here.
One thing I notice though: `progr_fun(number, number);` is not called any more.